En esta jupyter notebook se desarrolla la extracción, transformación y carga del conjunto de datos para 2019.

In [1]:
import pandas as pd
import json

%load_ext autoreload
%autoreload 2

import utils

import warnings
warnings.filterwarnings("ignore")

Se extraen los datos desde el archivo .csv, se convierte en Dataframe y se observa su contenido.

In [2]:
# Ruta al dataset
ruta_siniestro2021 = '../Datasets/feu-siniestros - (FC)feu-siniestros-2021.csv'

# Leer el archivo CSV directamente en un DataFrame
df_sin2021 = pd.read_csv(ruta_siniestro2021)

# Mostrar el DataFrame
df_sin2021

id_feu              unidad_regional siniestro_fecha siniestro_hora  \
0    128325  MUNICIPALIDAD DE CORRIENTES      2021-12-31       16:30:00   
1     93014            OBSERVATORIO VIAL      2021-12-29       15:15:00   
2    115166                      CAPITAL      2021-12-29        8:00:00   
3    134808            OBSERVATORIO VIAL      2021-12-29       16:30:00   
4     95909            OBSERVATORIO VIAL      2021-12-28        9:30:00   
..      ...                          ...             ...            ...   
650  132601  MUNICIPALIDAD DE CORRIENTES      2021-01-09        8:40:00   
651  108416            OBSERVATORIO VIAL      2021-01-07       21:36:00   
652  105919            OBSERVATORIO VIAL      2021-01-04       11:00:00   
653  108880                         O.V.      2021-01-02       22:00:00   
654  119957  MUNICIPALIDAD DE CORRIENTES      2021-01-01       10:50:00   

      provincia  id_provincia departamento  id_departamento   localidad  \
0    CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
1    CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
2    CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
3    CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
4    CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
..          ...           ...          ...              ...         ...   
650  CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
651  CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
652  CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
653  CORRIENTES            18      CAPITAL             21.0  CORRIENTES   
654  CORRIENTES            18      CAPITAL             21.0  CORRIENTES   

    zona_ocurrencia  ... senalizacion_carteleria senalizacion_horizontal  \
0            Urbana  ...                      No                      No   
1            Urbana  ...                      Si                      Si   
2            Urbana  ...                      Si                      Si   
3            Urbana  ...                      No                      No   
4            Urbana  ...                      Si                      Si   
..              ...  ...                     ...                     ...   
650          Urbana  ...                      Si                      Si   
651          Urbana  ...                      No                      No   
652          Urbana  ...                     S/D                     S/D   
653          Urbana  ...                     S/D                     S/D   
654          Urbana  ...                      Si                     NaN   

    senalizacion_transitoria luminosidad luz_artificial transito_restringido  \
0                         No         NaN             NO      Sin Restricción   
1                         No         Día             NO      Sin Restricción   
2                         No         Día             NO      Sin Restricción   
3                         No         Día             NO      Sin Restricción   
4                         No         Día             NO      Sin Restricción   
..                       ...         ...            ...                  ...   
650                       Si         Día             NO                  NaN   
651                       No       Noche             NO      Sin Restricción   
652                      S/D         Día             NO      Sin Restricción   
653                      S/D       Noche             NO      Sin Restricción   
654                      NaN         Día             NO      Sin Restricción   

    estado_ambiental  visibilidad  visibilidad_otro      semaforo  
0          Despejado        Buena               NaN  Sin Semáforo  
1          Despejado        Buena               NaN      Funciona  
2          Despejado        Buena               NaN      Funciona  
3          Despejado        Buena               NaN  Sin 

Se revisan los tipos de datos por columna y la cantidad de nulos.

In [3]:
utils.verificar_tipo_datos(df_sin2021)

nombre_campo                        tipo_datos  \
0                              id_feu                   [<class 'int'>]   
1                     unidad_regional                   [<class 'str'>]   
2                     siniestro_fecha                   [<class 'str'>]   
3                      siniestro_hora                   [<class 'str'>]   
4                           provincia                   [<class 'str'>]   
5                        id_provincia                   [<class 'int'>]   
6                        departamento  [<class 'str'>, <class 'float'>]   
7                     id_departamento                 [<class 'float'>]   
8                           localidad  [<class 'str'>, <class 'float'>]   
9                     zona_ocurrencia  [<class 'str'>, <class 'float'>]   
10                        via_publica  [<class 'str'>, <class 'float'>]   
11                         nombre_via  [<class 'str'>, <class 'float'>]   
12                          altura_km  [<class 'str'>, <class 'float'>]   
13                      entre_calle_1  [<class 'float'>, <class 'str'>]   
14                      entre_calle_2  [<class 'float'>, <class 'str'>]   
15                            latitud  [<class 'float'>, <class 'str'>]   
16                           longitud  [<class 'float'>, <class 'str'>]   
17                             ilesos                 [<class 'float'>]   
18                            heridos                   [<class 'int'>]   
19                         fallecidos                 [<class 'float'>]   
20                          vehiculos                   [<class 'int'>]   
21                           peatones                 [<class 'float'>]   
22            tipo_siniestro_multiple  [<class 'str'>, <class 'float'>]   
23               tipo_siniestro_unico  [<class 'str'>, <class 'float'>]   
24                    despiste_previo                   [<class 'str'>]   
25                tipo_siniestro_otro  [<class 'float'>, <class 'str'>]   
26                      tipo_colision  [<class 'str'>, <class 'float'>]   
27           cantidad_de_involucrados                   [<class 'int'>]   
28                        trazado_via  [<class 'str'>, <class 'float'>]   
29            configuracion_de_la_via  [<class 'str'>, <class 'float'>]   
30  configuracion_de_la_via_otro_tipo  [<class 'float'>, <class 'str'>]   
31                  pendiente_declive  [<class 'str'>, <class 'float'>]   
32                     cono_de_vision                 [<class 'float'>]   
33             material_de_la_calzada  [<class 'str'>, <class 'float'>]   
34        material_de_la_calzada_otro                 [<class 'float'>]   
35               estado_de_la_calzada  [<class 'str'>, <class 'float'>]   
36            estado_fisico_ambiental  [<class 'str'>, <class 'float'>]   
37        estado_de_conservacion_otro                 [<class 'float'>]   
38          division_fisica_de_la_via  [<class 'str'>, <class 'float'>]   
39     division_fisica_de_la_via_otro                 [<class 'float'>]   
40              barreras_de_seguridad                 [<class 'float'>]   
41                          banquinas                 [<class 'float'>]   
42            senalizacion_carteleria  [<class 'str'>, <class 'float'>]   
43            senalizacion_horizontal  [<class 'str'>, <class 'float'>]   
44           senalizacion_transitoria  [<class 'str'>, <class 'float'>]   
45                        luminosidad  [<class 'float'>, <class 'str'>]   
46                     luz_artificial  [<class 'str'>, <class 'float'>]   
47               transito_restringido  [<class 'str'>, <class 'float'>]   
48                   estado_ambiental  [<class 'str'>, <class 'float'>]   
49                        visibilidad  [<class 'str'>, <class 'float'>]   
50                   visibilidad_otro  [<class 'float'>, <class 'str'>]   
51                           semaforo  [<class 'str'>, <class 'float'>]   

    no_nulos_%  nulos_%  nulos  
0       100.00  

Se observa que hay columnas con más de 90% de nulos, por lo cual decidimos borrarlas.

In [4]:
# Eliminar columnas con más del 80% de valores nulos
umbral = 0.8  # Umbral del 80%
df_2021filtrado = df_sin2021.dropna(thresh=len(df_sin2021) * (1 - umbral), axis=1)

# Mostrar las columnas que se mantuvieron
df_2021filtrado.columns

Index(['id_feu', 'unidad_regional', 'siniestro_fecha', 'siniestro_hora',
       'provincia', 'id_provincia', 'departamento', 'id_departamento',
       'localidad', 'zona_ocurrencia', 'via_publica', 'nombre_via',
       'altura_km', 'entre_calle_1', 'latitud', 'longitud', 'ilesos',
       'heridos', 'fallecidos', 'vehiculos', 'peatones',
       'tipo_siniestro_multiple', 'tipo_siniestro_unico', 'despiste_previo',
       'tipo_colision', 'cantidad_de_involucrados', 'trazado_via',
       'configuracion_de_la_via', 'pendiente_declive',
       'material_de_la_calzada', 'estado_de_la_calzada',
       'estado_fisico_ambiental', 'division_fisica_de_la_via',
       'senalizacion_carteleria', 'senalizacion_horizontal',
       'senalizacion_transitoria', 'luminosidad', 'luz_artificial',
       'transito_restringido', 'estado_ambiental', 'visibilidad', 'semaforo'],
      dtype='object')

Verificamos ahora duplicados

In [5]:
utils.verifica_duplicados_por_columna(df_2021filtrado, 'id_feu')

'No hay duplicados'

No hay duplicados, podemos continuar corrigiendo los tipos de datos en algunas columnas. Por ejemplo, fechas y horas.

In [6]:
# Convertir la columna siniestro_fecha a tipo datetime
df_2021filtrado['siniestro_fecha'] = pd.to_datetime(df_2021filtrado['siniestro_fecha'], format='%Y-%m-%d', errors='coerce')

# Convertir la columna siniestro_hora a tipo time (si está en formato 24 horas)
df_2021filtrado['siniestro_hora'] = pd.to_datetime(df_2021filtrado['siniestro_hora'], format='%H:%M:%S', errors='coerce').dt.time

# Verificar las conversiones
df_2021filtrado[['siniestro_fecha', 'siniestro_hora']].head()

siniestro_fecha siniestro_hora
0      2021-12-31       16:30:00
1      2021-12-29       15:15:00
2      2021-12-29       08:00:00
3      2021-12-29       16:30:00
4      2021-12-28       09:30:00

Ahora, podemos pasar en limpio qué datos contiene este Dataframe

In [7]:
utils.verificar_tipo_datos(df_2021filtrado)

nombre_campo  \
0                      id_feu   
1             unidad_regional   
2             siniestro_fecha   
3              siniestro_hora   
4                   provincia   
5                id_provincia   
6                departamento   
7             id_departamento   
8                   localidad   
9             zona_ocurrencia   
10                via_publica   
11                 nombre_via   
12                  altura_km   
13              entre_calle_1   
14                    latitud   
15                   longitud   
16                     ilesos   
17                    heridos   
18                 fallecidos   
19                  vehiculos   
20                   peatones   
21    tipo_siniestro_multiple   
22       tipo_siniestro_unico   
23            despiste_previo   
24              tipo_colision   
25   cantidad_de_involucrados   
26                trazado_via   
27    configuracion_de_la_via   
28          pendiente_declive   
29     material_de_la_calzada   
30       estado_de_la_calzada   
31    estado_fisico_ambiental   
32  division_fisica_de_la_via   
33    senalizacion_carteleria   
34    senalizacion_horizontal   
35   senalizacion_transitoria   
36                luminosidad   
37             luz_artificial   
38       transito_restringido   
39           estado_ambiental   
40                visibilidad   
41                   semaforo   

                                           tipo_datos  no_nulos_%  nulos_%  \
0                                     [<class 'int'>]      100.00     0.00   
1                                     [<class 'str'>]      100.00     0.00   
2   [<class 'pandas._libs.tslibs.timestamps.Timest...      100.00     0.00   
3                           [<class 'datetime.time'>]      100.00     0.00   
4                                     [<class 'str'>]      100.00     0.00   
5                                     [<class 'int'>]      100.00     0.00   
6                    [<class 'str'>, <class 'float'>]       97.25     2.75   
7                                   [<class 'float'>]       97.25     2.75   
8                    [<class 'str'>, <class 'float'>]       98.02     1.98   
9                    [<class 'str'>, <class 'float'>]       99.39     0.61   
10                   [<class 'str'>, <class 'float'>]       99.54     0.46   
11                   [<class 'str'>, <class 'float'>]       98.17     1.83   
12                   [<class 'str'>, <class 'float'>]       81.68    18.32   
13                   [<class 'float'>, <class 'str'>]       75.73    24.27   
14                   [<class 'float'>, <class 'str'>]       48.40    51.60   
15                   [<class 'float'>, <class 'str'>]       48.40    51.60   
16                                  [<class 'float'>]       99.69     0.31   
17                                    [<class 'int'>]      100.00     0.00   
18                                  [<class 'float'>]       99.85     0.15   
19                                    [<class 'int'>]      100.00     0.00   
20                                  [<class 'float'>]       99.69     0.31   
21                   [<class 'str'>, <class 'float'>]       98.02     1.98   
22                   [<class 'str'>, <class 'float'>]       98.02     1.98   
23                                    [<class 'str'>]      100.00     0.00   
24                   [<class 'str'>, <class 'float'>]       36.95    63.05   
25                                    [<class 'int'>]      100.00     0.00   
26                   [<class 'str'>, <class 'float'>]       99.08     0.92   
27                   [<class 'str'>, <class 'float'>]       82.90    17.10   
28                   [<class 'str'>, <class 'float'>]       96.95     3.05   
29                   [<class 'str'>, <class 'float'>]       97.25     2.75   
30                   [<class 'str'>, <class 'float'>]       98.02     1.98   
31                   [<class 'str'>, <class 'float'>]       92.52     7.48   
32                   [<

In [8]:
print(df_2021filtrado.shape)

(655, 42)


Este conjunto contiene 42 columnas y quedó con 655 filas luego de borrar las columnas con mayor porcentaje de nulos. Las columnas que contiene son:

* id_feu: es el índice del siniestro.
* unidad_regional: es el nombre de la unidad regional encargada de cargar el dato. En este caso, el 100% corresponde a Observatorio Vial.
* siniestro_fecha: es la fecha del siniestro.
* siniestro_hora: es la hora del siniestro.
* id_provincia: es la provincia en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* departamento: es el departamento en el que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Capital.
* id_departamento: es la ID del departamento en el que ocurrió el siniestro.
* localidad: es la localidad en la que ocurrió el siniestro. En nuestro caso, el 100% corresponde a Corrientes.
* zona_ocurrencia: es la especificación sobre el tipo de zona donde ocurrió el siniestro. Urbana o Rural.
* via_publica: es la especificación sobre el tipo de calle en la que ocurrió el siniestro. Avenida, Calle o Ruta.
* nombre_via: es el nombre de la calle, avenida o ruta donde ocurrió el siniestro.
* altura_km: es la altura en kilómetros dónde ocurrió el siniestro.
* entre_calle_1: es el cruce en donde ocurrió el siniestro.
* latitud: es la latitud de dónde ocurrió el siniestro.
* longitud: es la longitud de dónde ocurrió el siniestro.
* ilesos: es la cantidad de personas que no sufrieron daños en el siniestro.
* heridos: es la cantidad de personas que sufrieron daños en el siniestro.
* fallecidos: es la cantidad de personas que fallecieron en el siniestro.
* vehiculos: es la cantidad de vehículos involucrados en el siniestro.
* peatones: es la cantidad de peatones involucrados en el siniestro.
* tipo_siniestro_multiple: es la especificación del hecho.
* despiste_previo: es la respuesta ante el hecho.
* tipo_colision: es el tipo de colisión del siniestro.
* cantidad_involucrados: es la cantidad de personas que participaron del siniestro.
* trazado_via: es la especficidad sobre el tipo de calle en que ocurrió el hecho. Recta o Curva.
* configuracion_de_la_via: es el tipo de configuración de la vía. Intersección, Cruce, Rotonda.
* pendiente_declive: es si existe pendiente.
* material_de_la_calzada: es el material del que está hecha la calzada donde ocurrió el siniestro.
* estado_de_la_calzada: es el estado en que se encontraba la calzada en el momento del siniestro.
* estado_fisico_ambiental: el estado de la calzada al momento del siniestro. Seca, Mojada.
* division_fisica_de_la_via: es el qué divide la vía donde ocurrió el siniestro. Por ejemplo, barrera de hormigón.
* senalización_carteleria: es si existía señalización en la zona al momento del hecho.
* senalización_horizontal: es si existía señalización horizontal en la zona al momento del hecho.
* senalización_transitoria: es si existía señalización transitoria en la zona al momento del hecho.
* luminosidad: es si el hecho ocurrió de día o de noche. Luz Natural.
* luz_artifical: es si en el lugar del hecho existía luz artificial.
* transito_restringido: es si el lugar era de tránsito restringido.
* estado_ambiental: es cómo estaba el tiempo al momento del hecho. Despejado, lluvioso, etc.
* visibilidad: es cómo se encontraba la visibilidad al momento del hecho. 
* semaforo: es si existe semáforo en la zona del hecho, y su funcionamiento.

In [9]:
# Eliminar columnas innecesarias
columnas_a_eliminar = [
    'unidad_regional',
    'id_provincia',
    'id_departamento',
    'altura_km',
    'tipo_siniestro_multiple',
    'despiste_previo',
    'tipo_siniestro_unico',
    'trazado_via',
    'configuracion_de_la_via',
    'pendiente_declive',
    'division_fisica_de_la_via',
    'senalizacion_carteleria',
    'senalizacion_horizontal',
    'senalizacion_transitoria',
    'transito_restringido', 
    'provincia',
    'departamento',
    'localidad',
    ]

df_2021filtrado = df_2021filtrado.drop(columns=columnas_a_eliminar)


Ahora tenemos un Dataframe mucho más manejable, con los datos más importantes para nuestra investigación. Vamos a guardarlo para su futuro uso.

In [10]:
sin2021 = '../Datasets_limpios/siniestros2021.csv'
df_2021filtrado.to_csv(sin2021)